src/App.tsx

```diff
@@ -15,36 +16,113 @@ export const goodsFromServer = [
+const SORT_FIELD_LENGTH = 'length';
+
+function getPreparedGoods(
+  goods: string[],
+  { sortType, isReverse }: { sortType: string; isReverse: boolean },
+) {
+  const preparedGoods = [...goods];
+
+  if (sortType) {
+    preparedGoods.sort((good1, good2) => {
+      switch (sortType) {
+        case SORT_FIELD_ALPH:
+          return good1.localeCompare(good2);
+
+        case SORT_FIELD_LENGTH:
+          return good1.length - good2.length;
+
+        default:
+          return 0;
+      }
+    });
+  }
+
+  if (isReverse) {
+    preparedGoods.reverse();
+  }
+
+  return preparedGoods;
+}
+
 export const App: React.FC = () => {
+  const [sortField, setSortField] = useState('');
+  const [isReverse, setIsReverse] = useState(false);
+
+  const handleReset = () => {
+    setSortField('');
+    setIsReverse(false);
+  };
+
+  const handleSortByAlphabet = () => {
+    setSortField(SORT_FIELD_ALPH);
+  };
+
+  const handleSortByLength = () => {
+    setSortField(SORT_FIELD_LENGTH);
+  };
+
+  const handleReverse = () => {
+    setIsReverse(prevState => !prevState);
+  };
+
+  const visibleGoods = getPreparedGoods(goodsFromServer, {
+    sortType: sortField,
+    isReverse,
+  });
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          onClick={handleSortByAlphabet}
+          type="button"
+          className={cn('button is-info', {
+            'is-light': sortField !== SORT_FIELD_ALPH,
+          })}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          onClick={handleSortByLength}
+          type="button"
+          className={cn('button is-success', {
+            'is-light': sortField !== SORT_FIELD_LENGTH,
+          })}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          onClick={handleReverse}
+          type="button"
+          className={cn('button is-warning', {
+            'is-light': !isReverse,
+          })}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {(sortField || isReverse) && (
+          <button
+            onClick={handleReset}
+            type="button"
+            className="button is-danger is-light"
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {visibleGoods.map(good => (
+          <li data-cy="Good" key={good}>
```

Comment:
Using the good's name as a `key` is not recommended if the names are not guaranteed to be unique. It would be better to use an identifier like an index or a unique ID for each item when mapping over an array to generate a list of React elements.